src/js/module.js

```diff
@@ -8,3 +8,43 @@ Module.noImageDecoding = true;
 Module.noAudioDecoding = true;
 Module.noWasmDecoding = false; // we preload wasm using the built in plugin now
 Module.preloadedWasm = {};
+
+/**
+ *
+ * @param {undefined|(() => string)} stdin
+ * @param {undefined|((text: string) => void)} stdout
+ * @param {undefined|((text: string) => void)} stderr
+ */
+export function setStandardStreams(stdin, stdout, stderr) {
```

Comment:
How about this for the function body:
```
    let input = [];
    let inputIndex = 0;
    let encoder = new TextEncoder();
    function stdin() {
      if(inputIndex < input.length){
        let character = input[inputIndex];
	    inputIndex++;
	    return character;
      }
      let newInput = config.input();
      if(newInput === undefined){
        return null;
      } 
      if(typeof newInput !== "string"){
      	throw new TypeError(`stdin callback returned invalid output type ${typeof newInput}, expected a string or undefined.`);
      }
      input = encoder.encode(newInput + "\n\x00");
      inputIndex = 1;
      return input[0] || null; // we need to return a `null` to terminate the string.
    }
```

Reply:
Seems nice! 
One minor issue that I have with it is that this would make inputting zero bytes into one line of `stdin` impossible. Sure, nobody is ever going to do that, but it feels weird using zero bytes as terminators for strings.

Reply:
I guess this could hypothetically be harmful if the string had actual null bytes in it, but I suspect such null bytes will be interpreted by C to terminate the string anyways.

Reply:
I will look at the internals and see if they pass the length on to the C APIs.

Reply:
I decided to just test it out by hardcoding the user input to `text = "cat\0dog";`

Apparently the null bytes really do get preserved. I'm pretty surprised at the fact that emscripten actually handles that
![image](https://user-images.githubusercontent.com/10220080/126895088-f2feb7dc-aa9b-463e-9483-44c178238f9d.png)

Reply:
Hmm it does look like without this null bytes would be handled correctly since `read` returns how many bytes were written to the buffer. So better to keep the approach that handles null bytes just in case (they are allowed to appear both in js strings and in Python strings, so no reason we should convert them wrong).